### PR TITLE
feat: add benchmark

### PR DESCRIPTION
**Pull request by Lampese**

### CODE REVIEW

**Comment by lijunchen** [comment]
Our initial plan was to provide `moon bench` and use the `"bench"` label to identify a test block for benchmarking.

```
test "xxx" "bench" {
  f()
}
```

Since this feature hasn't been implemented yet, this alternative looks good to me.

**Comment by peter-jerry-ye** [comment]
Is it possible to add js backend support as well? Otherwise we'll have to split the CI

**Comment by Lampese** [comment]
> Is it possible to add js backend support as well? Otherwise we'll have to split the CI

Let me try it.

File: benchmark/benchmark.mbt [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+  name : String
+  f : () -> Unit
+  count : Int
+}
+
+struct TaskResult {
+  task : Task
+  average : Double
+  max : Double
+  min : Double
```
Comment (peter-jerry-ye) [inline benchmark/benchmark.mbt:25]: Are these fields expected to be private?

File: benchmark/benchmark.mbt [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  $|Min Time Per Test: \{self.min}s/per test
+  $|----------------------------------------
+}
+
+pub fn output(self : TaskResult, logger : Logger) -> Unit {
+  logger.write_string(self.to_string())
+}
+
+pub struct Criterion {
+  tasks : Array[Task]
```
Comment (peter-jerry-ye) [inline benchmark/benchmark.mbt:64]: Is this field expected to be public?

Comment (Lampese) [reply benchmark/benchmark.mbt:64]: No, maybe priv is better.

File: benchmark/benchmark.mbt (hunk 1 above)
Comment (Lampese) [reply benchmark/benchmark.mbt]: No, maybe pub is better.

**Review by peter-jerry-ye — APPROVED** [review]
LGTM. Don't forget to update the interfaces.

**Comment by Lampese** [comment]
Why does workflow here need approval? Core doesn't need any approval, but it runs more CI than here everyday.

**Comment by peter-jerry-ye** [comment]
Because you're the first-time contributor here. https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-workflow-runs/approving-workflow-runs-from-public-forks

**Comment by Lampese** [comment]
> Because you're the first-time contributor here. https://docs.github.com/en/actions/managing-workflow-runs-and-deployments/managing-workflow-runs/approving-workflow-runs-from-public-forks

Allright :)

File: benchmark/benchmark.mbt (hunk 1 above)
Comment (Lampese) [reply benchmark/benchmark.mbt]: I realized something, it seems that these fields can also be accessed when TaskResult is not set to pub.

**Review by lijunchen — APPROVED** [review]

**Comment by Lampese** [comment]
I pushed a new commit to match the first review by petter-jerry-ye.

**Comment by peter-jerry-ye** [comment]
@Lampese To make it clear: Do you expect people to get the `max` `min` `average` from the `TaskResult`? Would you like to write a white box test to demonstrate your expected usage?

**Comment by Lampese** [comment]
> @Lampese To make it clear: Do you expect people to get the `max` `min` `average` from the `TaskResult`? Would you like to write a white box test to demonstrate your expected usage?

Yeah, I think it's expected behavior, for instance, if our user get a TaskResult, they not just can use its Show trait to output it and they can use the benchmark result to do more thing I think.

```moonbit
let criterion = Criterion::new()
criterion.add(Task::new("sum", fn() { sum(10000000) |> ignore }, count=100))
let result = criterion.run()["sum"]
println(result.average) // like this
if result.average < xxx {
  //do more thing
}
```

**Comment by peter-jerry-ye** [comment]
Then I think you'll need to declare it as `pub` or at least `pub(readonly)`. Verify this with a white box testing would be helpful.

**Comment by Lampese** [comment]
> Then I think you'll need to declare it as `pub` or at least `pub(readonly)`. Verify this with a white box testing would be helpful.

Oh I know what you mean, I will do that.

**Pull request merged**